### PR TITLE
Shutdown and drain gRPC completion queue on exit

### PR DESCRIPTION
**Pull request by McKayJT**
This prevents a segmentation fault in next() when it tries to call back
to a deleted object when murmur is shut down.

### CODE REVIEW

File: src/murmur/MurmurGRPCImpl.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ namespace MurmurRPC {
 class MurmurRPCImpl : public QThread {
 		Q_OBJECT;
 		std::unique_ptr<grpc::Server> m_server;
+		volatile bool b_IsRunning;
```
Comment (hacst) [inline src/murmur/MurmurGRPCImpl.h]: Should be m_isRunning (see https://wiki.mumble.info/wiki/Coding_Guidelines#Member_variables)

File: src/murmur/MurmurGRPCImpl.cpp [hunk 2]
```diff
@@ -138,10 +140,18 @@ MurmurRPCImpl::MurmurRPCImpl(const QString &address, std::shared_ptr<::grpc::Ser
...
 	start();
 }
 
 MurmurRPCImpl::~MurmurRPCImpl() {
+	void *ignored_tag;
+	bool ignored_ok;
+	b_IsRunning = false;
+	m_server->Shutdown(std::chrono::system_clock::now());
+	m_completionQueue->Shutdown();
+	while (m_completionQueue->Next(&ignored_tag, &ignored_ok))
```
Comment (hacst) [inline src/murmur/MurmurGRPCImpl.cpp]: Should delete the drained tags instead of leaking them. No reason to needlessly create valgrind/asan false positives even if it doesn't endanger application stability.

File: src/murmur/MurmurGRPCImpl.cpp [hunk 3]
```diff
@@ -138,10 +140,18 @@ MurmurRPCImpl::MurmurRPCImpl(const QString &address, std::shared_ptr<::grpc::Ser
...
 	m_server = builder.BuildAndStart();
 	meta->connectListener(this);
+	b_IsRunning = true;
 	start();
 }
 
 MurmurRPCImpl::~MurmurRPCImpl() {
+	void *ignored_tag;
+	bool ignored_ok;
+	b_IsRunning = false;
```
Comment (hacst) [inline src/murmur/MurmurGRPCImpl.cpp]: Even with this ::run might still pick up another tag and start processing it or it might be processing one right now. Then the destructor might return before processing in ::run is complete.

TBO looking at the shutdown sequence I'm wondering whether it is still safe to process anything at all at the point of this destructor or whether grpc shutdown would have to happen before server shutdowns to be safe. Did you happen to investigate that?

Comment (McKayJT) [reply src/murmur/MurmurGRPCImpl.cpp]: I did investigate a little about this. Yes, there is a problem here, that I decided not to fix as the server is shutting down anyway.

Say you have something listening for events using the streaming protocols. If you you set the timeout to something like 10s in the future, call stop on the server, and don't stop the main loop (hoping to drain it there) it invokes the oom killer. If you don't set a deadline, and try to wait for the server to finish with m_server->Wait(), it hangs. Nothing is really able to be completed at this point in the process. You can also call m_server->Wait() after requesting the shutdown, it just causes it to hang until the deadline hits and everything is disconnected.

TL;DR: Yes, it will try to do one more action, but that action will fail because the server shutdown has been requested and the deadline is in the past.

File: src/murmur/MurmurGRPCImpl.cpp (hunk 2 above)
Comment (McKayJT) [reply src/murmur/MurmurGRPCImpl.cpp]: I can delete these. I didn't because the server is about to be shut down and everything is going to be freed anyway.

File: src/murmur/MurmurGRPCImpl.h (hunk 1 above)
Comment (McKayJT) [reply src/murmur/MurmurGRPCImpl.h]: This is what I get for trying to remember the code style guidelines off the top of my head.

**Comment by Kissaki** [comment]
Typo in commit message title “shudown”

**Comment by hacst** [comment]
LGTM

**Comment by Kissaki** [comment]
The commit message of 1d2e5b090af1a02e7bbdf8bf3bbf679635834645 was not fixed.
